[2d] unflatten_tensor on compute stream for DTensorExtension

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116678
* #116573
* __->__ #116559
* #116426

Context: Existing FSDPExtension have some bug in the case when the
unflatten tensor involves some compute/communications in cuda stream,
the current logic of FSDPExtension unflatten tensor happens in the
unshard stream, which makes runtime lost sync with the compute stream,
and if there're some dependencies between the compute stream and the
unflatten tensor logic, currently it would lose sync point, which could
possibly lead to NaN.

This PR make the FSDPExtension to record the compute stream and let
DTensorExtension to directly use the compute stream for unflatten_tensor.

In long term we might want to directly make the FSDP runtime logic to only
make the unshard happen in unshard stream, and use unshard views to
happen in the compute stream. We currently fix this in the Extension
directly as this is the simplest thing to do without affecting FSDP
runtime logic

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[2d] unflatten_tensor on compute stream for DTensorExtension

Context: Existing FSDPExtension have some bug in the case when the
unflatten tensor involves some compute/communications in cuda stream,
the current logic of FSDPExtension unflatten tensor happens in the
unshard stream, which makes runtime lost sync with the compute stream,
and if there're some dependencies between the compute stream and the
unflatten tensor logic, currently it would lose sync point, which could
possibly lead to NaN.

This PR make the FSDPExtension to record the compute stream and let
DTensorExtension to directly use the compute stream for unflatten_tensor.

In long term we might want to directly make the FSDP runtime logic to only
make the unshard happen in unshard stream, and use unshard views to
happen in the compute stream. We currently fix this in the Extension
directly as this is the simplest thing to do without affecting FSDP
runtime logic

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] unflatten_tensor on compute stream for DTensorExtension"


Context: Existing FSDPExtension have some bug in the case when the
unflatten tensor involves some compute/communications in cuda stream,
the current logic of FSDPExtension unflatten tensor happens in the
unshard stream, which makes runtime lost sync with the compute stream,
and if there're some dependencies between the compute stream and the
unflatten tensor logic, currently it would lose sync point, which could
possibly lead to NaN.

This PR make the FSDPExtension to record the compute stream and let
DTensorExtension to directly use the compute stream for unflatten_tensor.

In long term we might want to directly make the FSDP runtime logic to only
make the unshard happen in unshard stream, and use unshard views to
happen in the compute stream. We currently fix this in the Extension
directly as this is the simplest thing to do without affecting FSDP
runtime logic

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented:
> In long term we might want to directly make the FSDP runtime logic to only
make the unshard happen in unshard stream, and use unshard views to
happen in the compute stream. We currently fix this in the Extension
directly as this is the simplest thing to do without affecting FSDP
runtime logic

cc: @weifengpy This was what I was talking about with regard to reorganizing the logical "unshard" and "wait" ops as an approach to the forward prefetching fix.

awgu commented on torch/distributed/tensor/parallel/fsdp.py:
```diff
@@ -320,19 +320,48 @@ class DTensorExtensions(FSDPExtensions):
+            # need to sync properly.
+            # TODO: this is a short term fix and we should make the get_unflat_views
+            # directly happen in the compute stream.
+            with self.device_handle.stream(self.compute_stream):
```
nit: If you wanted to avoid repeating the core code:
```
stream = self.compute_stream or torch.cuda.current_stream()
with self.device_handle.stream(stream):
    ...
```
(Overhead of entering the stream context does not matter if it only runs during construction time.)

awgu reviewed (approved):
The change makes sense to me!

fduwjj commented on test/distributed/fsdp/test_fsdp_tp_integration.py:
```diff
@@ -373,6 +373,14 @@ def forward(self, x):
+        grads = [p.grad for p in fsdp_2d_model.parameters() if p.grad is not None]
+
+        for grad in grads:
+            if isinstance(grad, DTensor):
+                grad = grad.full_tensor()
```
In the current 2D design, do we see a case when a gradient is DTensor? They are all expected to be flattened tensors IIUC. Curious to learn how this is possible?

wanchaol replied:
Oh yes that's right! I forgot to update the test case, will update soon

yifuwang reviewed (approved):
Nice fix!

We probably want to make accidental DTensor stream drifting impossible or easier to detect. Not in this diff though.

wanchaol committed:
Update on "[2d] unflatten_tensor on compute stream for DTensorExtension"


Context: Existing FSDPExtension have some bug in the case when the
unflatten tensor involves some compute/communications in cuda stream,
the current logic of FSDPExtension unflatten tensor happens in the
unshard stream, which makes runtime lost sync with the compute stream,
and if there're some dependencies between the compute stream and the
unflatten tensor logic, currently it would lose sync point, which could
possibly lead to NaN.

This PR make the FSDPExtension to record the compute stream and let
DTensorExtension to directly use the compute stream for unflatten_tensor.

In long term we might want to directly make the FSDP runtime logic to only
make the unshard happen in unshard stream, and use unshard views to
happen in the compute stream. We currently fix this in the Extension
directly as this is the simplest thing to do without affecting FSDP
runtime logic

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

bdhirsh commented on torch/distributed/tensor/parallel/fsdp.py:
```diff
@@ -320,19 +320,36 @@ class DTensorExtensions(FSDPExtensions):
     ) -> Tuple[torch.Tensor, Optional[Any]]:
         return _flatten_tensor(tensor)
 
+    @torch._dynamo.disable
```
Will this matter for @voznesenskym 's work tracing FSDP 🤔

wanchaol replied:
I tried remove this and dynamo complains about a stream related issue. I guess once @voznesenskym enable the tracing FSDP we can visit this to ensure the 2D parallelism don't trigger graph break
